### ANN

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import datetime
import matplotlib.pyplot as plt
import seaborn as sns
import utils
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LeakyReLU, PReLU, ELU
from keras.layers import Dropout

In [1]:
from keras import backend as K
def root_mean_squared_error(y_true, y_pred):
        return K.sqrt(K.mean(K.square(y_pred - y_true)))

Using TensorFlow backend.


In [3]:
df_train = pd.read_csv('dataset/df_train.csv')
df_test = pd.read_csv('dataset/df_test.csv')
target = df_train['SalePrice']
df_train = df_train.drop(['SalePrice'], axis = 1)
X_train, X_test, y_train, y_test = train_test_split(df_train, target, test_size = 0.25, random_state = 42)

In [4]:
#initializing the ANN
classifier = Sequential()

#adding the layers
classifier.add(Dense(units = 50, kernel_initializer='he_uniform', activation='relu', input_dim = X_train.shape[1]))

classifier.add(Dense(units = 25, kernel_initializer='he_uniform', activation='relu'))

classifier.add(Dense(units = 50, kernel_initializer='he_uniform', activation='relu'))

classifier.add(Dense(units = 1, kernel_initializer='he_uniform'))

#compile the ANN
classifier.compile(loss = root_mean_squared_error, optimizer = 'Adamax')

#fitting the ann to the training set
model_history = classifier.fit(X_train.values, y_train.values, validation_split=0.20, batch_size=10, nb_epoch=1000)





Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:17: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.



Train on 874 samples, validate on 219 samples
Epoch 1/1000
874/874 [==============================] - 0s 459us/step - loss: 4.4844 - val_loss: 1.8538
Epoch 2/1000
874/874 [==============================] - 0s 133us/step - loss: 2.0617 - val_loss: 3.5058
Epoch 3/1000
874/874 [==============================] - 0s 133us/step - loss: 1.7994 - val_loss: 2.8834
Epoch 4/1000
874/874 [==============================] - 0s 115us/step - loss: 2.0017 - val_loss: 2.5315
Epoch 5/1000
874/874 [==============================] - 0s 115us/step - loss: 2.2299 - val_loss: 7.5555
Epoch 6/1000
874/874 [==============================] - 0s 115us/step - loss: 3.2504 - val_loss: 4.5565
Epoch 7/1000
874/874 [==============================] - 0s 115us/step - loss: 2.5051 - val_loss: 4.9599
Epoch 8/1000
874/874 [==============================] - 0s 115us/step - loss: 2.3902 - val_loss: 3.0148
Epoch 9/1000
874/874 [==============================] - 0s 115us/step - loss: 1.3139 - val_loss: 1.7859
Epoch 10/1000
874

Epoch 77/1000
874/874 [==============================] - 0s 125us/step - loss: 0.3055 - val_loss: 0.3282
Epoch 78/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2691 - val_loss: 0.2556
Epoch 79/1000
874/874 [==============================] - 0s 125us/step - loss: 0.3741 - val_loss: 0.6531
Epoch 80/1000
874/874 [==============================] - 0s 107us/step - loss: 0.4971 - val_loss: 0.1774
Epoch 81/1000
874/874 [==============================] - 0s 125us/step - loss: 0.3936 - val_loss: 0.4263
Epoch 82/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2718 - val_loss: 0.4714
Epoch 83/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2865 - val_loss: 0.1779
Epoch 84/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2424 - val_loss: 0.5354
Epoch 85/1000
874/874 [==============================] - 0s 107us/step - loss: 0.4622 - val_loss: 0.3613
Epoch 86/1000
874/874 [==============================] 

874/874 [==============================] - 0s 125us/step - loss: 0.3388 - val_loss: 0.2158
Epoch 155/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2141 - val_loss: 0.1763
Epoch 156/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2925 - val_loss: 0.1452
Epoch 157/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2353 - val_loss: 0.5871
Epoch 158/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2895 - val_loss: 0.4368
Epoch 159/1000
874/874 [==============================] - 0s 143us/step - loss: 0.2507 - val_loss: 0.3462
Epoch 160/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2444 - val_loss: 0.2069
Epoch 161/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2868 - val_loss: 0.4701
Epoch 162/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2516 - val_loss: 0.4059
Epoch 163/1000
874/874 [==============================] - 0s 

874/874 [==============================] - 0s 125us/step - loss: 0.2581 - val_loss: 0.2002
Epoch 232/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2846 - val_loss: 0.1809
Epoch 233/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2028 - val_loss: 0.2634
Epoch 234/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2498 - val_loss: 0.3651
Epoch 235/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2090 - val_loss: 0.1719
Epoch 236/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2097 - val_loss: 0.2033
Epoch 237/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2315 - val_loss: 0.2990
Epoch 238/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2244 - val_loss: 0.8351
Epoch 239/1000
874/874 [==============================] - 0s 125us/step - loss: 0.3264 - val_loss: 0.5127
Epoch 240/1000
874/874 [==============================] - 0s 

874/874 [==============================] - 0s 131us/step - loss: 0.2117 - val_loss: 0.2547
Epoch 309/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2160 - val_loss: 0.2417
Epoch 310/1000
874/874 [==============================] - 0s 125us/step - loss: 0.3001 - val_loss: 0.2213
Epoch 311/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2516 - val_loss: 0.4055
Epoch 312/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2523 - val_loss: 0.2951
Epoch 313/1000
874/874 [==============================] - 0s 125us/step - loss: 0.1772 - val_loss: 0.3831
Epoch 314/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2268 - val_loss: 0.2413
Epoch 315/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2361 - val_loss: 0.2425
Epoch 316/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2640 - val_loss: 0.7937
Epoch 317/1000
874/874 [==============================] - 0s 

874/874 [==============================] - 0s 107us/step - loss: 0.1707 - val_loss: 0.2618
Epoch 386/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2202 - val_loss: 0.3569
Epoch 387/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2360 - val_loss: 0.3069
Epoch 388/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2092 - val_loss: 0.2253
Epoch 389/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2186 - val_loss: 0.3690
Epoch 390/1000
874/874 [==============================] - 0s 107us/step - loss: 0.1920 - val_loss: 0.1766
Epoch 391/1000
874/874 [==============================] - 0s 125us/step - loss: 0.2764 - val_loss: 0.1471
Epoch 392/1000
874/874 [==============================] - 0s 107us/step - loss: 0.2179 - val_loss: 0.2594
Epoch 393/1000
874/874 [==============================] - 0s 89us/step - loss: 0.2134 - val_loss: 0.1492
Epoch 394/1000
874/874 [==============================] - 0s 8

874/874 [==============================] - 0s 188us/step - loss: 0.1748 - val_loss: 0.2328
Epoch 463/1000
874/874 [==============================] - 0s 185us/step - loss: 0.1808 - val_loss: 0.1962
Epoch 464/1000
874/874 [==============================] - 0s 185us/step - loss: 0.1784 - val_loss: 0.1298
Epoch 465/1000
874/874 [==============================] - 0s 187us/step - loss: 0.2436 - val_loss: 0.1943
Epoch 466/1000
874/874 [==============================] - 0s 208us/step - loss: 0.1862 - val_loss: 0.1641
Epoch 467/1000
874/874 [==============================] - 0s 185us/step - loss: 0.2038 - val_loss: 0.1383
Epoch 468/1000
874/874 [==============================] - 0s 159us/step - loss: 0.2056 - val_loss: 0.3352
Epoch 469/1000
874/874 [==============================] - 0s 159us/step - loss: 0.2065 - val_loss: 0.1295
Epoch 470/1000
874/874 [==============================] - 0s 162us/step - loss: 0.1821 - val_loss: 0.1833
Epoch 471/1000
874/874 [==============================] - 0s 

874/874 [==============================] - 0s 115us/step - loss: 0.1528 - val_loss: 0.3692
Epoch 540/1000
874/874 [==============================] - 0s 115us/step - loss: 0.2267 - val_loss: 0.1858
Epoch 541/1000
874/874 [==============================] - 0s 97us/step - loss: 0.1845 - val_loss: 0.1393
Epoch 542/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1701 - val_loss: 0.1282
Epoch 543/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1437 - val_loss: 0.1286
Epoch 544/1000
874/874 [==============================] - 0s 97us/step - loss: 0.1977 - val_loss: 0.2307
Epoch 545/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1390 - val_loss: 0.1303
Epoch 546/1000
874/874 [==============================] - 0s 97us/step - loss: 0.1713 - val_loss: 0.2915
Epoch 547/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1963 - val_loss: 0.1694
Epoch 548/1000
874/874 [==============================] - 0s 115

874/874 [==============================] - 0s 115us/step - loss: 0.1643 - val_loss: 0.1578
Epoch 617/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1590 - val_loss: 0.1288
Epoch 618/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1617 - val_loss: 0.1265
Epoch 619/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1534 - val_loss: 0.1265
Epoch 620/1000
874/874 [==============================] - 0s 133us/step - loss: 0.1602 - val_loss: 0.3268
Epoch 621/1000
874/874 [==============================] - 0s 142us/step - loss: 0.1432 - val_loss: 0.1573
Epoch 622/1000
874/874 [==============================] - 0s 150us/step - loss: 0.1596 - val_loss: 0.2200
Epoch 623/1000
874/874 [==============================] - 0s 176us/step - loss: 0.1718 - val_loss: 0.1291
Epoch 624/1000
874/874 [==============================] - 0s 150us/step - loss: 0.1534 - val_loss: 0.1266
Epoch 625/1000
874/874 [==============================] - 0s 

874/874 [==============================] - 0s 115us/step - loss: 0.1962 - val_loss: 0.2202
Epoch 694/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1469 - val_loss: 0.1333
Epoch 695/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1606 - val_loss: 0.1782
Epoch 696/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1581 - val_loss: 0.1428
Epoch 697/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1342 - val_loss: 0.1461
Epoch 698/1000
874/874 [==============================] - 0s 133us/step - loss: 0.1500 - val_loss: 0.1273
Epoch 699/1000
874/874 [==============================] - 0s 192us/step - loss: 0.1407 - val_loss: 0.1613
Epoch 700/1000
874/874 [==============================] - 0s 194us/step - loss: 0.1389 - val_loss: 0.2873
Epoch 701/1000
874/874 [==============================] - 0s 144us/step - loss: 0.1460 - val_loss: 0.1578
Epoch 702/1000
874/874 [==============================] - 0s 

874/874 [==============================] - 0s 167us/step - loss: 0.2103 - val_loss: 0.1990
Epoch 771/1000
874/874 [==============================] - 0s 145us/step - loss: 0.1422 - val_loss: 0.1314
Epoch 772/1000
874/874 [==============================] - 0s 158us/step - loss: 0.1324 - val_loss: 0.1460
Epoch 773/1000
874/874 [==============================] - 0s 150us/step - loss: 0.1394 - val_loss: 0.1395
Epoch 774/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1560 - val_loss: 0.1259
Epoch 775/1000
874/874 [==============================] - 0s 133us/step - loss: 0.1595 - val_loss: 0.1683
Epoch 776/1000
874/874 [==============================] - 0s 133us/step - loss: 0.1404 - val_loss: 0.1431
Epoch 777/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1708 - val_loss: 0.2134
Epoch 778/1000
874/874 [==============================] - 0s 97us/step - loss: 0.2136 - val_loss: 0.3335
Epoch 779/1000
874/874 [==============================] - 0s 9

874/874 [==============================] - 0s 158us/step - loss: 0.1369 - val_loss: 0.1335
Epoch 848/1000
874/874 [==============================] - 0s 150us/step - loss: 0.1389 - val_loss: 0.1560
Epoch 849/1000
874/874 [==============================] - 0s 178us/step - loss: 0.1234 - val_loss: 0.1603
Epoch 850/1000
874/874 [==============================] - 0s 168us/step - loss: 0.1494 - val_loss: 0.1285
Epoch 851/1000
874/874 [==============================] - 0s 194us/step - loss: 0.1273 - val_loss: 0.1339
Epoch 852/1000
874/874 [==============================] - 0s 283us/step - loss: 0.1446 - val_loss: 0.1268
Epoch 853/1000
874/874 [==============================] - 0s 196us/step - loss: 0.1417 - val_loss: 0.1740
Epoch 854/1000
874/874 [==============================] - 0s 173us/step - loss: 0.1524 - val_loss: 0.1324
Epoch 855/1000
874/874 [==============================] - 0s 173us/step - loss: 0.1242 - val_loss: 0.1273
Epoch 856/1000
874/874 [==============================] - 0s 

874/874 [==============================] - 0s 115us/step - loss: 0.1389 - val_loss: 0.1551
Epoch 925/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1239 - val_loss: 0.3326
Epoch 926/1000
874/874 [==============================] - 0s 97us/step - loss: 0.1437 - val_loss: 0.1681
Epoch 927/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1408 - val_loss: 0.1489
Epoch 928/1000
874/874 [==============================] - 0s 89us/step - loss: 0.1236 - val_loss: 0.1288
Epoch 929/1000
874/874 [==============================] - 0s 97us/step - loss: 0.1208 - val_loss: 0.1619
Epoch 930/1000
874/874 [==============================] - 0s 97us/step - loss: 0.1267 - val_loss: 0.1258
Epoch 931/1000
874/874 [==============================] - 0s 115us/step - loss: 0.1332 - val_loss: 0.1349
Epoch 932/1000
874/874 [==============================] - 0s 97us/step - loss: 0.1431 - val_loss: 0.1487
Epoch 933/1000
874/874 [==============================] - 0s 178us

In [26]:
ann_pred = classifier.predict(np.array(X_test))
r2, rmse, rmsle = utils.evaluate_model(ann_pred.reshape(-1), y_test)
df_metric = pd.DataFrame({'model':'ANN','rmse':[rmse], 'r2':[r2], 'rmsle':[rmsle]})

In [28]:
df_metric = df_metric.set_index('model')
df_metric.head()

rmse       r2    rmsle
model                            
ANN    0.148397  0.86457  0.01161

The model seems to perform well with the r2 value of 0.86 and rmse and rmsle value as mentioned above